datasource_generic.go

```diff
@@ -0,0 +1,130 @@
+package gapi
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+)
+
+type DataSourceGeneric struct {
+	Id     int64  `json:"id,omitempty"`
+	Name   string `json:"name"`
+	Type   string `json:"type"`
+	URL    string `json:"url"`
+	Access string `json:"access"`
+
+	Database string `json:"database,omitempty"`
+	User     string `json:"user,omitempty"`
+	// Deprecated in favor of secureJsonData.password
+	Password string `json:"password,omitempty"`
+
+	OrgId     int64 `json:"orgId,omitempty"`
+	IsDefault bool  `json:"isDefault"`
+
+	BasicAuth     bool   `json:"basicAuth"`
+	BasicAuthUser string `json:"basicAuthUser,omitempty"`
+	// Deprecated in favor of secureJsonData.basicAuthPassword
+	BasicAuthPassword string `json:"basicAuthPassword,omitempty"`
+
+	JSONData       JsonData `json:"jsonData,omitempty"`
+	SecureJSONData JsonData `json:"secureJsonData,omitempty"`
+}
+
+type JsonData map[string]interface{}
+
+func (c *Client) NewDataSourceGeneric(s *DataSourceGeneric) (int64, error) {
+	data, err := json.Marshal(s)
+	if err != nil {
+		return 0, err
+	}
+	req, err := c.newRequest("POST", "/api/datasources", nil, bytes.NewBuffer(data))
+	if err != nil {
+		return 0, err
+	}
+
+	resp, err := c.Do(req)
+	if err != nil {
+		return 0, err
+	}
+	if resp.StatusCode != 200 {
```

Comment:
I think all 2xx status codes should be OK:
```suggestion
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
```